### ISSUE TITLE
Are empty/nil slices externalized correctly?

### ISSUE THREAD

**Issue by shurcooL** [gopherjs/gopherjs#399]
I know that `(interface{})(nil)` gets externalized as a JavaScript `null`, and `js.Undefined` gets externalized as JavaScript `undefined`.

I also know that, according to the [conversion table](https://godoc.org/github.com/gopherjs/gopherjs/js), `[]byte` slices get externalized as `Uint8Array`. But other slices, e.g., `[]string`, get externalized as `Array`.

However, it seems that it's not always the case. Empty `[]byte` slices seemingly get externalized as JavaScript `Array` rather than `Uint8Array`.

``` bash
$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", (interface{})(nil))'
null

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", js.Undefined)'
undefined

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []byte{1, 2, 3})'
Uint8Array { '0': 1, '1': 2, '2': 3 }

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []byte{})'
[]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", ([]byte)(nil))'
[]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []string{"hi", "bye"})'
[ 'hi', 'bye' ]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []string{})'
[]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", ([]string)(nil))'
[]
```

Is this intentional by design, or an accidental oversight?

I've run into this as part of https://github.com/goxjs/gl/issues/18.

**Comment by shurcooL** [comment]
The reason I ask is because I would expect `s.Global.Get("console").Call("log", []byte{})` to get externalized as an empty `Uint8Array`:

```
Uint8Array {}
```

Not an empty `Array`.

Or is there a reason why a numeric slice with 0 elements is treated differently than a numeric slice with 1 or more elements?

**Comment by shurcooL** [comment]
> Is this intentional by design, or an accidental oversight?

I'm guessing this was an accidental oversight, given that #597 looks like similar/identical issue, and it was resolved by #599.

Running the same code as above with latest GopherJS produces different results, that look more in line with what's expected:

```
$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", (interface{})(nil))'
null

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", js.Undefined)'
undefined

$  goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []byte{1, 2, 3})'
Uint8Array [ 1, 2, 3 ]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []byte{})'
Uint8Array [  ]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", ([]byte)(nil))'
Uint8Array [  ]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []string{"hi", "bye"})'
[ 'hi', 'bye' ]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", []string{})'
[]

$ goexec -quiet -compiler=gopherjs 'js.Global.Get("console").Call("log", ([]string)(nil))'
[]
```

**Comment by shurcooL** [comment]
> I'm guessing this was an accidental oversight, given that #597 (/cc @flimzy) looks like similar/identical issue, and it was resolved by #599.

Since no one has disputed my guess above so far, I'm going to take that to be the reality.

Closing as duplicate of #597.

**Comment by myitcv** [comment]
@shurcooL apologies I missed this for some reason. 

I'll make sure I look at this in https://github.com/gopherjs/gopherjs/issues/617#issuecomment-297129541

**Comment by shurcooL** [comment]
> 4\. Ensure we address #399

@myitcv, #597 (the issue this is a duplicate of) has already been resolved by #599.
